File: app/src/main/java/wangdaye/com/geometricweather/common/basic/models/options/unit/PollenUnit.java
package wangdaye.com.geometricweather.common.basic.models.options.unit;

import android.content.Context;
import android.text.BidiFormatter;

import androidx.annotation.Nullable;

import wangdaye.com.geometricweather.R;
import wangdaye.com.geometricweather.common.utils.DisplayUtils;

public enum PollenUnit {

    PPCM("ppcm", 0);

    private final String unitId;
    private final int unitArrayIndex;

    PollenUnit(String id, int arrayIndex) {
        unitId = id;
        unitArrayIndex = arrayIndex;
    }

    public String getUnitId() {
        return unitId;
    }

    public String getPollenText(Context context, @Nullable Integer value) {
        if (value == null) {
            return getPollenText(context, 0);
        } else {
            return getPollenText(context, (int) value);
        }
    }

    public String getPollenText(Context context, int value) {
        return getPollenText(context, value, DisplayUtils.isRtl(context));
    }

    private String getPollenText(Context context, int value, boolean rtl) {
        if (rtl) {
            return BidiFormatter.getInstance().unicodeWrap(UnitUtils.formatInt(value))
                    + context.getResources().getStringArray(R.array.pollen_units)[unitArrayIndex];
        } else {
            return UnitUtils.formatInt(value)
                    + context.getResources().getStringArray(R.array.pollen_units)[unitArrayIndex];
        }
    }

    public String getPollenVoice(Context context, @Nullable Integer value) {
        if (value == null) {
            return getPollenVoice(context, 0);
        } else {
            return getPollenVoice(context, (int) value);
        }
    }

    public String getPollenVoice(Context context, int value) {
        return getPollenVoice(context, value, DisplayUtils.isRtl(context));
    }

    private String getPollenVoice(Context context, int value, boolean rtl) {
        if (rtl) {
            return BidiFormatter.getInstance().unicodeWrap(UnitUtils.formatInt(value))
                    + context.getResources().getStringArray(R.array.pollen_unit_voices)[unitArrayIndex];
        } else {
            return UnitUtils.formatInt(value)
                    + context.getResources().getStringArray(R.array.pollen_unit_voices)[unitArrayIndex];
        }
    }
}

File: app/src/main/java/wangdaye/com/geometricweather/common/basic/models/options/unit/PollenUnit.kt
package wangdaye.com.geometricweather.common.basic.models.options.unit

import android.content.Context
import wangdaye.com.geometricweather.R
import wangdaye.com.geometricweather.common.basic.models.options._basic.UnitEnum
import wangdaye.com.geometricweather.common.basic.models.options._basic.Utils
import wangdaye.com.geometricweather.common.utils.DisplayUtils

enum class PollenUnit(
    override val id: String,
    override val unitFactor: Float
): UnitEnum<Int> {

    PPCM("ppcm", 1f);

    override val valueArrayId = R.array.pollen_unit_values
    override val nameArrayId = R.array.pollen_units
    override val voiceArrayId = R.array.pollen_unit_voices

    override fun getName(context: Context) = Utils.getName(context, this)

    override fun getVoice(context: Context) = Utils.getVoice(context, this)

    override fun getValueWithoutUnit(
        valueInDefaultUnit: Int
    ) = (valueInDefaultUnit * unitFactor).toInt()

    override fun getValueInDefaultUnit(
        valueInCurrentUnit: Int
    ) = (valueInCurrentUnit / unitFactor).toInt()

    override fun getValueTextWithoutUnit(
        valueInDefaultUnit: Int
    ) = Utils.getValueTextWithoutUnit(this, valueInDefaultUnit)!!

    override fun getValueText(
        context: Context,
        valueInDefaultUnit: Int
    ) = getValueText(context, valueInDefaultUnit, DisplayUtils.isRtl(context))

    override fun getValueText(
        context: Context,
        valueInDefaultUnit: Int,
        rtl: Boolean
    ) = Utils.getValueText(
        context = context,
        enum = this,
        valueInDefaultUnit = valueInDefaultUnit,
        rtl = rtl
    )

    override fun getValueVoice(
        context: Context,
        valueInDefaultUnit: Int
    ) = getValueVoice(context, valueInDefaultUnit, DisplayUtils.isRtl(context))

    override fun getValueVoice(
        context: Context,
        valueInDefaultUnit: Int,
        rtl: Boolean
    ) = Utils.getVoiceText(
        context = context,
        enum = this,
        valueInDefaultUnit = valueInDefaultUnit,
        rtl = rtl
    )
}